### PR TITLE
Fixed login with password and better configuration file handling

### PR DESCRIPTION
**Pull request by siredmar**
Fixed login with password (if not passed) and better configuration file handling.

**Test case 1: no config file existing, no username given**
```
$ ls ~/.mender-clirc
ls: cannot access '/home/armin/.mender-clirc': No such file or directory
$ ./mender-cli login 
FAILURE: No username set
```

**Test case 2: no config file existing, correct username given**
```
$ ls ~/.mender-clirc
ls: cannot access '/home/armin/.mender-clirc': No such file or directory

$ ./mender-cli login --username user@example.com
Password:************
login successful
```

**Test case 3: no config file existing, incorrect username given**
```
$ ls ~/.mender-clirc
ls: cannot access '/home/armin/.mender-clirc': No such file or directory
$ ./mender-cli login --username user@example.com
Password:************
FAILURE: login failed with status 401
```

**Test case 4: config file exists, correct username and pw**
```
./mender-cli login
Using configuration file: /home/armin/.mender-clirc
login successful
```

**Test case 5: config file exists, invalid json file**
```
./mender-cli login
Failed to read config: While parsing config: invalid character ':' after top-level value
```

**Test case 7: config file exists, incorrect username**
```
 ./mender-cli login
Using configuration file: /home/armin/.mender-clirc
FAILURE: login failed with status 401
```

**Test case 8: config file exists, correct username, incorrect pw**
```
uration file: /home/armin/.mender-clirc
FAILURE: login failed with status 401
```

**Test case 9: config file exists, correct username, pw missing**
```
./mender-cli login
Using configuration file: /home/armin/.mender-clirc
Password:******************************
login successful
```

Changelog: None
Signed-off-by: Armin Schlegel <armin.schlegel@gmx.de>

### CODE REVIEW

File: cmd/login.go [hunk 1]
```diff
@@ -56,9 +56,11 @@ func init() {
 	viper.AddConfigPath("/etc/mender-cli/")
 	viper.AddConfigPath("$HOME/")
 	viper.AddConfigPath(".")
-	err := viper.ReadInConfig()
-	if err != nil {
-		log.Info("No configuration file found")
+	if err := viper.ReadInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); !ok {
+			log.Info(fmt.Sprintf("Failed to read config: %s", err))
```
Comment (oleorhagen) [inline cmd/login.go:61]: Suggestion:
```suggestion
			log.Infof("Failed to read config: %s", err.Error())
```

File: cmd/login.go [hunk 2]
```diff
@@ -56,9 +56,11 @@ func init() {
...
+			os.Exit(1)
```
Comment (oleorhagen) [inline cmd/login.go:62]: Question:
Maybe we should log the error in this instance also?

File: cmd/root.go [hunk 3]
```diff
@@ -65,7 +65,7 @@ func init() {
 	// Here you will define your flags and configuration settings.
 	// Cobra supports persistent flags, which, if defined here,
 	// will be global for your application.
-	rootCmd.PersistentFlags().StringP(argRootServer, "", "https://hosted.mender.io", "root server URL, e.g. 'https://hosted.mender.io' (default)")
+	rootCmd.PersistentFlags().StringP(argRootServer, "", "https://hosted.mender.io", "root server URL, e.g. 'https://hosted.mender.io'")
```
Comment (oleorhagen) [inline cmd/root.go:68]: Question:
Not really related to the commit message, no?

File: cmd/login.go [hunk 4]
```diff
@@ -93,9 +95,6 @@ func NewLoginCmd(cmd *cobra.Command, args []string) (*LoginCmd, error) {
 	}
 
 	password := viper.GetString(argLoginPassword)
-	if password == "" {
```
Comment (oleorhagen) [inline cmd/login.go:96]: Makes sense, thx.

File: cmd/login.go [hunk 5]
```diff
@@ -56,9 +56,11 @@ func init() {
 	viper.AddConfigPath("/etc/mender-cli/")
 	viper.AddConfigPath("$HOME/")
 	viper.AddConfigPath(".")
-	err := viper.ReadInConfig()
-	if err != nil {
-		log.Info("No configuration file found")
+	if err := viper.ReadInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); !ok {
```
Comment (oleorhagen) [inline cmd/login.go:60]: Suggestion:
Import the `errors` package, and do:
```
if errors.Is(err, viper.ConfigFileNotFoundError) {
...
```

**Review by oleorhagen — COMMENTED** [review]
Real nice!

Your choices make sense, I have left some minor suggestions and nitpicks. But all in all, this looks real good.

Great PR description, always nice.

It might even be time to actually start adding some tests to the tool at this point. I understand if you feel this is beyond the scope though.

Well done!

File: cmd/login.go (hunk 1 above)
Comment (siredmar) [reply cmd/login.go:61]: Infof does not exist in log.go

File: cmd/login.go (hunk 2 above)
Comment (siredmar) [reply cmd/login.go:62]: I don't see this as an error, because the parameters can also be set as command line args.

File: cmd/root.go (hunk 3 above)
Comment (siredmar) [reply cmd/root.go:68]: You are right, but useless. I saw it and changed it.

File: cmd/login.go (hunk 5 above)
Comment (siredmar) [reply cmd/login.go:60]: `type viper.ConfigFileNotFoundError is not an expression`. ConfigFileNotFoundError is basically is not of type error.

**Comment by siredmar** [comment]
@oleorhagen please see conversations. If you are ok, please merge. 
I have another PR in the pipeline that adds a `artifacts list` feature that i will put when this PR is merged.

File: cmd/login.go (hunk 1 above)
Comment (oleorhagen) [reply cmd/login.go:61]: Ur, right! 
Surprised me a little.

File: cmd/login.go (hunk 5 above)
Comment (oleorhagen) [reply cmd/login.go:60]: `viper.ConfigFileNotFoundError` is a struct. So it needs the:
```
errors.Is(err, viper.ConfigFileNotFoundError{})
```

**Review by oleorhagen — CHANGES_REQUESTED** [review]
> @oleorhagen please see conversations. If you are ok, please merge.
> I have another PR in the pipeline that adds a `artifacts list` feature that i will put when this PR is merged.

Real nice. I still have some comments left on the PR though (Please don't close conversations until they are resolved).

It is fine to open the other PR's anyways. We will just rebase them on top of master after this is merged )

File: cmd/root.go (hunk 3 above)
Comment (siredmar) [reply cmd/root.go:68]: No, the "(default") as is was shwon double

File: cmd/login.go (hunk 5 above)
Comment (siredmar) [reply cmd/login.go:60]: You are right! Thx!

File: cmd/login.go (hunk 2 above)
Comment (siredmar) [reply cmd/login.go:62]: My initial implementation was checking both cases resulting in a silent continuation if the config file was not pressent. I will add a log.Info in this case.

File: cmd/root.go (hunk 3 above)
Comment (siredmar) [reply cmd/root.go:68]: Only to have mentioned it again explicitly:
```
$ ./mender-cli --help

...
Flags:
  -h, --help            help for mender-cli
      --server string   root server URL, e.g. 'https://hosted.mender.io (default)' (default "https://hosted.mender.io")
```

File: cmd/login.go (hunk 5 above)
Comment (siredmar) [reply cmd/login.go:60]: First errors.Is internally compares if they are comparable (check if type compares). If true, it checks if `err == viper.ConfigFileNotFoundError{}`. Problem is that `err` contains data and the empty struct for `ConfigFileNotFoundError` doesn't this check won't succeed - ever.

Comment (oleorhagen) [reply cmd/login.go:60]: Ahh, I see. Alright, let's leave it as it is then

**Review by oleorhagen — APPROVED** [review]

**Pull request merged**